[자동차 경주 게임] 미션 제출합니다

devholic22 opened the pull request:
`기능을 구현하기 전에 README.md 파일에 구현할 기능 목록을 정리해 추가한다.`를 진행했습니다.

* `Application`에서 `GameController`를 실행하여 게임을 시작하도록 했습니다.
* 출력과 관련된 로직은 `Printer`에서 이루어지도록 했습니다.
* 랜덤 값을 구하는 로직은 `Dice`에서 이루어지도록 했습니다.
* 게임 예시 진행을 봤을 때, 목표 진행 횟수까지 전진된 `Car`가 발생될 때 까지 진행해야 하는 것 같아 이 방식으로 게임이 진행되도록 했습니다.
* 테스트 코드를 아직 작성하지 못했습니다. 일단은 테스트 코드 전에 기본적인 코드 습관에 대해 먼저 교정받고 싶은데 그렇게 해도 괜찮을까요?

devholic22 committed:
docs: README.md 작성

devholic22 committed:
feat: 1차 기능 개발

sosow0212 commented on README.md:
```diff
@@ -5,3 +5,24 @@
 
 - [온라인 코드 리뷰 과정](https://github.com/woowacourse/woowacourse-docs/blob/master/maincourse/README.md)
+
+## 구현할 기능 목록
```
구현이 완료 됐다면 check도 해주세요!

sosow0212 commented on README.md:
```diff
@@ -5,3 +5,24 @@
 ## 우아한테크코스 코드리뷰
```
++ 커밋 단위가 너무 큽니다.

기능 단위 별로 커밋 날려주십숑

sosow0212 replied:
++ 테스트도 추가해주세요

sosow0212 commented on src/main/java/Car.java:
```diff
@@ -0,0 +1,49 @@
+
+    private Car(String name) {
+        this.name = name;
+        this.distance = 0;
```
매직 넘버는 상수처리해서 의미를 부여해주세요.

0은 통상적으로 기본 값이지만, 만약 5부터 시작된다면 다른 사람이 이해할 수 있을까요!

devholic22 replied:
클린 코드에도 작성 (387p)되어 있는 내용이군요! 반영했습니다 :)

sosow0212 commented on src/main/java/Car.java:
```diff
@@ -0,0 +1,49 @@
+        this.distance = 0;
+    }
+
+    public static Car of(String name) {
```
정적 팩토리 네이밍은 암묵적으로 사용하지만, `from`, `of`의 차이점은 무엇일까요?🤔

devholic22 replied:
from은 하나의 매개변수를 받아 객체를 생성, of는 여러 개의 매개변수를 받아 객체를 생성하는군요 :) 하나와 여러 개일 때 사용하는 것은 알았지만 이 둘을 사용할 때의 시점을 헷갈렸던 것 같습니다 🥲

sosow0212 replied:
이 부분은 저도 헷갈려 했던 부분인데 금방 적응될 것 같아요 ㅎㅎ
반면에 꼭 from, of라는 네이밍을 고집할 필요 없이 의미를 줘도 됩니다! (그게 정팩메의 장점이니깐요)

예를 들면 저는 초기 상태를 저장해주는 것이 목표라면 `from or of`보다는 `createDefault`와 같은 이름을 부여해주는 편이기도 합니다!

sosow0212 commented on src/main/java/Car.java:
```diff
@@ -0,0 +1,49 @@
+        return new Car(name);
+    }
+
+    public static boolean isValueSatisfiedToAccelerate(int value) {
+        final int ACCELERATE_MINIMUM_VALUE = 4;
+        return value >= ACCELERATE_MINIMUM_VALUE;
+    }
```
이런 구조로 만드신 이유가 궁금합니다.

```
if() {
    return true;
}

return false;
```

해당 구조로 변경해도 될 것 같습니다.
그리고 `ACCELERATE_MINIMUM_VALUE`는 상수 같은데 따로 빼주시는게 좋아보여요

devholic22 replied:
* 메서드 상수를 클래스 상수로 변경 완료했습니다!
* 처음에 `value >= ACCERATE_MINIMUM_VALUE`로 작성한 이유는 if문을 작성하지 않고도 바로 판단할 수 있기 때문이었는데, 생각해보니 더 명시적으로 나타내기 위해서는 if문을 걸어두는 게 좋은 것 같네요..!
* 또 가끔씩 if문 다음에 바로 return이 나올 경우에는 중괄호를 작성하지 않곤 했었는데, 구글 컨벤션을 보니 생략하는 경우가 아예 없도록 하고 있군요! 수정 완료했습니다 :)

sosow0212 replied:
위와 같이 리뷰드린 이유는 `ACCELERATE_MINIMUM_VALUE`라는 네이밍으로 메서드 내에서의 상수로 관리하시는 것 같은데, 클래스 레벨에서 상수로 관리하면 어떨까 싶어서 리뷰드린 내용입니다!

사실 상수로 뺀다면 `return ...`로 바로 끝내셔도 됩니다 ㅎㅎ
이 부분과 같이 간단한 메서드는 메서드 네이밍만 잘 정해주시면 뜻이 전달되기에, 굳이 명시적으로 하실 필요는 없어보여요

sosow0212 commented on src/main/java/Car.java:
```diff
@@ -0,0 +1,49 @@
+        return value >= ACCELERATE_MINIMUM_VALUE;
+    }
+
+    public String getName() {
```
getter, setter 메서드는 맨 밑으로!

devholic22 replied:
반영했습니다 :) 그런데 [스택오버플로우 글](https://stackoverflow.com/questions/4668218/are-there-any-java-method-ordering-conventions)을 보면, getter & setter 메서드를 맨 밑으로 작성하는게 따로 공식적으로 존재하는 컨벤션은 아닌 것 같은데 (`Not sure if...`) 이렇게 하는 이유가 있을까요?

sosow0212 replied:
맞습니다 사실 딱 정해져있는 컨벤션은 없지만, 저도 예전에 관련 리뷰를 받은 적이 있었습니다.
그 당시 getter 같은 부분은 쓰는 일이 많고 안 쓸 수 있다면 지양하는 편이 좋기 때문에 명시적으로 아래에 위치하고, 불필요한 getter는 없애기 위해서 아래쪽에서 사용하는 것이 좋다고 들었습니다 :)

sosow0212 commented on src/main/java/Car.java:
```diff
@@ -0,0 +1,49 @@
+        return this.distance;
+    }
+
+    public String answerDistance() {
```
Car 객체가 View에 의존하는 것 같습니다.
다른 방법으로 View에게 메시지를 전달하는 방법은 없을까요?

객체지향의 사실과오해 책을 읽어보시면 좋을 것 같아요

devholic22 replied:
출력을 담당하는 Printer가 차의 정보를 출력하도록 변경하였습니다. (`printCarDistanceStatus`)
그런데 Name 객체 (원시값 포장)를 생성하다보니, Car에서 `answerStatus` 메서드를 실행할 때 이름을 가져오기 위해서는 참조하고 있는 Name 객체로부터 `getName` 메서드를 호출해야만 실제 이름을 가져오게 되는데 이런 경우는 저번에 리뷰해주신 `메서드일 시 내부적인 로직이 추가될 경우 영향을 받을 수 있다`는 것에 충돌될 것 같다는 생각이 듭니다. 혹시 이 부분을 개선할 방법이 있을지 궁금합니다.
```java
// Car에서 answerStatus를 호출할 경우
public void answerStatus() {
    Printer.printCarDistanceStatus(name.getName(), START_DISTANCE, distance);
}
```

sosow0212 replied:
질문을 잘 이해한지 모르겠습니다.
`Car` 객체 안에  `Name` 객체를 원시값 포장을 했다는 뜻이죠?

그렇다면 `car.getName()` 메서드 안의 로직은 `car.getName().getName() -> Name 객체의 getName() 구조`가 되겠죠?

> 메서드일 시 내부적인 로직이 추가될 경우 영향을 받을 수 있다
이 리뷰와는 별개의 내용 같습니다!

도메인이 뷰에 의존하는 것을 없애려면
뷰의 매개변수에서 `car`객체 자체가 넘어가는 것보다는 매개변수로 `String name`과 같은 값을 넘겨주면 되지 않을까요? 도메인 상위 레이어인 외부에서 전달해주는 것이니깐요.

추가적으로 dto를 우리가 사용하는 이유와 어떻게 사용하는지를 생각해보면 좋을 것 같아요. (지금 dto를 쓰라는 뜻은 아닙니다!)

물론 지금 하시는 게 스프링은 아니지만, 주로 우리가 스프링으로 API 서버를 만들 때 클라이언트 단에게 객체 자체를 주기보다는 순수 값으로 주는 경우가 많죠? 똑같이 해당 미션에서도 View 영역은 클라이언트 단이라고 생각해보시면 편하게 이해하실 수 있을 것 같아요

sosow0212 commented on src/main/java/Car.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    public boolean isDistanceEqualTo(int distance) {
+        return getDistance() == distance;
```
getter가 크게 바뀔 일은 없겠지만, 만약 로직이 추가되면 어떻게 될까요? 다른 메서드에 의존하기 보다는 `this.value == distance`와 같이 반환해주는게 좋아보입니다

devholic22 replied:
메서드다보니 `내부적인 로직이 추가될 경우 영향을 받을 수 있다`는 사실을 간과하고 있었네요..! 수정했습니다 :)

sosow0212 commented on src/main/java/Dice.java:
```diff
@@ -0,0 +1,7 @@
+public class Dice {
+
+    public static int getRandomValue() {
+        final int DICE_MAX_VALUE = 10;
```
상수로 따로 빼주세요

devholic22 replied:
적용 완료하였습니다!

sosow0212 commented on src/main/java/GameController.java:
```diff
@@ -0,0 +1,95 @@
+
+public class GameController {
+
+    private static BufferedReader br;
```
1. 변수명에서 생략은 피해주세요 `br`이 어떤 건지 협업하는 사람을 알 수 없을 것 같아요~
더 명확한 이름을 주시는 게 클린코드 관점에서 더 좋을 것 같습니다.

2. `static final`, `final`, `static`을 각각 언제 쓸까요?

devholic22 replied:
> 1. 변수명에서 생략은 피해주세요 `br`이 어떤 건지 협업하는 사람을 알 수 없을 것 같아요~
>    더 명확한 이름을 주시는 게 클린코드 관점에서 더 좋을 것 같습니다.
> 2. `static final`, `final`, `static`을 각각 언제 쓸까요?

* 1번: `br`을 `inputReader`로 변경 완료하였습니다!
* 2번: `static`은 `객체 생성 없이 사용할 수 있는 필드와 메소드를 생성`하고자 할 때, `final`은 `값이 저장된 후 변경할 수 없도록` 할 때, `static final`은 `완전히 값이 달라질 수 없는 상수를 사용`하고자 할 때 (final만 쓰면 객체마다 값이 달라질 수 있으므로) 사용하는 것을 다시 알게 되었습니다. `GameController`의 경우에는 `BufferedReader`를 사용하는 게 한 번으로 고정되고 객체 생성 없이 사용하는 케이스가 없으므로 `BufferedReader`에 `static`을 해제하고 `final`을 부여하도록 변경하였습니다! 알고리즘 문제들을 풀을 때 매번 `BufferedReader`를 `private static final`로 사용했었는데 제대로 의미를 모른 채 사용했었음을 알게 되었습니다 :)

sosow0212 commented on src/main/java/GameController.java:
```diff
@@ -0,0 +1,95 @@
+public class GameController {
+
+    private static BufferedReader br;
+    private static List<Car> playCars;
```
`List<Car> playCars`로 플레이어들의 상태를 관리하는 것으로 보입니다.

`일급 컬렉션`에 대해서 공부해보시는 건 어떨까요?
[일급 컬렉션 보러가기](https://blog.naver.com/sosow0212/223014652755)

devholic22 replied:
구현 완료하였습니다 감사합니다 :)

sosow0212 commented on src/main/java/GameController.java:
```diff
@@ -0,0 +1,95 @@
+        close();
+    }
+
+    private void createCars(String input) {
```
일급 컬렉션을 공부하고 이 부분을 객체에게 위임해보는 건 어떨까요?

sosow0212 replied:
++ `String[] carNames`와 같이 배열로 말고 List로 받는다면 stream을 이용해서 해당 코드를 많이 줄일 수 있을 것 같습니다.

devholic22 replied:
일급 컬렉션인 Cars를 구현하여 적용 완료하였습니다! stream이 진짜 이렇게까지 획기적일 수 있구나를 깨닫게 되었습니다 🥹

sosow0212 commented on src/main/java/GameController.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    private boolean isWinnerNotExist() {
+        return playCars.stream().noneMatch(car -> car.isDistanceEqualTo(playDistance));
```
구글에 검색 하셔서 `객체지향 생활체조 4번 규칙` 참고해주세요~

devholic22 replied:
적용 완료하였습니다! 그런데 혹시 `.filter(car -> car.method())`와 같은 상황처럼 점이 두 개 붙은 상황도 점을 내려야 하는지 궁금합니다 👀

sosow0212 replied:
괄호 안에 저런 점은 보통 내리진 않습니다.. ㅋㅋ
물론 가독성에 저해 된다면 팀원과 컨벤션을 만들고 내려도 되겠지만 일반적으로 제 주변에선 개행을 추가해주진 않는 것 같습니다.

sosow0212 commented on src/main/java/GameController.java:
```diff
@@ -0,0 +1,95 @@
+        this.playDistance = Integer.parseInt(input);
+    }
+
+    private boolean validateName(String name) {
```
이 부분도 객체에게 위임해야겠죠?

sosow0212 replied:
`원시값 포장`에 대해서 공부해보면 좋을 것 같습니다.
[원시값 포장 보러가기](https://blog.naver.com/sosow0212/223025314498)

devholic22 replied:
말씀하신 내용들을 Name에 위임하여 구현 완료하였습니다! 확실히 원시값 Name을 포장해두니 Name에 관련된 이름 검증 등을 GameController가 관여하지 않게 되어 객체지향적으로 설계된 것 같습니다 👍

sosow0212 commented on src/main/java/GameController.java:
```diff
@@ -0,0 +1,95 @@
+
+    private boolean isUsedName(String name) {
+        for (Car car : playCars) {
+            if (car.getName().equals(name))
```
`car.getName().equals(...)`보다는 car 객체에게 일을 시키는게 좋지 않을까요?
이 역시 객체지향의 사실과오해 책 읽어보시면 이해가 되실 것 같아요 ㅎㅎ

sosow0212 replied:
요구사항 중에 인덴트가 1개만 있어야하는데 for -> if로 인덴트가 2이네요 확인해주세요~

여러가지 방법이 있지만 추천 키워드를 드리자면 `stream` 학습해보시면 좋을 것 같습니다 !

devholic22 replied:
> `car.getName().equals(...)`보다는 car 객체에게 일을 시키는게 좋지 않을까요? 이 역시 객체지향의 사실과오해 책 읽어보시면 이해가 되실 것 같아요 ㅎㅎ

적용 완료하였습니다!

sosow0212 commented on src/main/java/Printer.java:
```diff
@@ -0,0 +1,47 @@
+    private static boolean isOnlyOneWinner(List<String> winners) {
+        return winners.size() == 1;
+    }
+}
```
[해당 글](https://yeonyeon.tistory.com/241)을 확인해주세요~

devholic22 replied:
적용 완료하였습니다!

sosow0212 reviewed (changes requested):
작업하시느라 수고하셨습니다.
남긴 코멘트 몇 가지 확인해주시고 반영해주세요~
추가적으로 테스트가 누락되어 있는데 테스트도 같이 작성해주세요.

devholic22 committed:
refactor: 매직 넘버 (distance) 상수 처리

devholic22 committed:
docs: README.md 기능 체크

devholic22 committed:
refactor: 정적 팩터리 메서드 이름 교정

devholic22 committed:
refactor: getter 메서드 위치 변경

devholic22 committed:
refactor: isValueSatisfiedToAccelerate 메서드 표현 변경 및 메서드 상수를 클래스 상수로 변경

devholic22 committed:
refactor: isDistanceEqualTo 메서드 수정

devholic22 committed:
refactor: Car에게 이름 묻도록 변경

devholic22 committed:
refactor: 이름 길이 메서드 상수를 클래스 상수로 변경

devholic22 committed:
refactor: BufferedReader 이름 변경

devholic22 committed:
refactor: BufferedReader static 해제 및 final 부여

devholic22 committed:
refactor: DICE_MAX_VALUE 클래스 상수로 변경

devholic22 committed:
refactor: Name 원시값 포장 생성

devholic22 committed:
refactor: Car가 View 기능을 수행하지 않도록 변경

devholic22 committed:
refactor: 이름 길이 검증을 Name에게 옮김

devholic22 committed:
refactor: RANDOM_RANGE_EXPAND 상수 정의

devholic22 committed:
refactor: Name 분리 책임을 GameController에서 Name으로 이동

devholic22 committed:
refactor: nameCandidates -> candidateNames 변수 이름 변경

devholic22 committed:
refactor: Name에서 Name 리스트 생성 및 중복처리를 하도록 변경

devholic22 committed:
refactor: 일급 컬렉션 (Cars) 생성 및 책임 위임

devholic22 committed:
refactor: Car에서 필요없는 메서드 삭제

devholic22 committed:
refactor: Name에서 필요없는 메서드 삭제

devholic22 committed:
fix: createNameList 공백 제거 버그 수정

devholic22 committed:
test: 정상적인 이름 리스트 입력 시 모두 저장되어야 한다.

devholic22 committed:
fix: 이름 입력 중복 처리 버그 수정 (EqualsAndHashCode 작성)

devholic22 committed:
test: 중복 이름이 있다면 한 번만 저장되어야 한다.

devholic22 committed:
test: 이름은 1자 이상 5자 이하여야만 저장된다.

devholic22 committed:
refactor: 중복 코드 삭제 (name 공백 제거)

devholic22 committed:
test: 빈 문자열 입력 시 빈 리스트를 반환해야 한다.

devholic22 committed:
test: 공백 이름이 입력된 경우 무시되어야 한다.

devholic22 committed:
test: null 입력이 주어진 경우 예외가 발생해야 한다.

devholic22 committed:
test: Name 객체의 getName 메서드가 정상적으로 동작해야 한다.

devholic22 committed:
test: 이름을 토대로 자동차가 생성되어야 한다.

devholic22 committed:
이름이 잘못되었다면 자동차 자체가 생성이 되지 않아야 한다.

devholic22 committed:
test: 처음 진행된 거리는 0이어야 한다.

devholic22 committed:
refactor: getDistance 삭제

devholic22 committed:
test: 값이 문제가 없다면 거리는 1번 전진되어야 한다.

devholic22 committed:
feat: 값이 문제가 있다면 거리는 전진되지 않아야 한다.

devholic22 committed:
docs: distanceIncreaseTest 문구 수정

devholic22 committed:
test: 값이 최소 조건 값을 충족하면 거리는 전진되지 않아야 한다.

devholic22 committed:
test: 값이 최대 조건 값을 초과할 경우 거리는 전진되지 않아야 한다.

devholic22 committed:
test: Dice (랜덤값) 테스트 using 전략 패턴 (Strategy pattern)

- 전략 패턴 구현 규칙에 따라, Dice를 인터페이스화
- CarTest에서 랜덤값과 관련된 테스트를 전부 DiceTest로 옮김
- 테스트 시 사용할 수 있는 케이스에 해당되는 Dice 구현체들을 만듦
- 랜덤 값 계산 방식 수정

sosow0212 commented on src/main/java/Car.java:
```diff
@@ -34,8 +34,9 @@ public boolean isDistanceEqualTo(int distance) {
-        if (Car.isValueSatisfiedToAccelerate(dice)) {
+    public void racing(Dice dice) {
+        final int randomNumber = dice.random();
+        if (Car.isValueSatisfiedToAccelerate(randomNumber)) {
```
`Car` 객체 내부적으로 존재하는 static 메서드를 사용한 이유가 있으신가요?
다른 방법으로 할 수 있을 것 같아요!

devholic22 replied:
Car 객체에서 Car static 메서드를 사용했다니.. 비효율적이었네요 😅 수정했습니다!

sosow0212 commented on src/main/java/Car.java:
```diff
@@ -0,0 +1,49 @@
+    private static final int START_DISTANCE = 0;
+    private static final int ACCELERATE_MINIMUM_VALUE = 4;
+    private static final int ACCELERATE_MAXIMUM_VALUE = 9;
+    private final Name name;
```
상수와 객체 사이 개행 하나 주세요~

devholic22 replied:
수정했습니다! 인텔리제이의 `사용 위치`에 가려져서 실수했었네요 🥲

sosow0212 commented on src/main/java/Car.java:
```diff
@@ -0,0 +1,49 @@
+    private final Name name;
+    private int distance;
+
+    private Car(Name name) {
```
`Car`의 생성자는 `Car(final Name name, final int distance)`로 받고
정적 팩토리 메서드에서 `static Car from(final Name name)`으로 두는 건 어떨까요?

조금 더 유연하게 사용할 수 있을 것 같습니다

devholic22 replied:
수정했습니다!

sosow0212 commented on src/main/java/Car.java:
```diff
@@ -0,0 +1,49 @@
+        this.distance++;
+    }
+
+    public void answerStatus() {
+        Printer.printCarDistanceStatus(name.getName(), START_DISTANCE, distance);
+    }
```
도메인이 뷰를 알고 있어야할까요?
저번 리뷰에 이어서 도메인과 뷰는 최대한 의존하면 안됩니다.

차라리 도메인에서 getter를 이용해서 컨트롤러에서 값을 꺼내고 뷰에 넘겨주는 방식으로 의존을 끊어내는 것은 어떨까요?

devholic22 replied:
수정했습니다!

devholic22 replied:
음.. 그런데 Cars에서 Printer를 호출하는 것은 어쩔 수 없는 걸까요? 각 car들이 레이싱 한 후 Printer에게 출력하라고 하는 것이다보니 Cars에서 Printer를 호출하는 게 남아있는 것 같습니다.

sosow0212 replied:
Cars에서 Printer를 호출한다는 뜻인가요?

도메인이랑 뷰랑 연결되어 있는데 도메인은 뷰를 알아야할까요? 반대로 뷰가 도메인을 알아야할까요?

그런 방법을 안쓰고 도메인에서의 getter를 이용해 값을 꺼내고 컨트롤러를 통해 뷰에게 넘어간다면 도메인을 몰라도 뷰에선 작업을 할 수 있을 것 같습니다

sosow0212 commented on src/main/java/Cars.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class Cars {
```
일급컬렉션 👍

sosow0212 commented on src/main/java/Cars.java:
```diff
@@ -0,0 +1,31 @@
+    public boolean hasNotWinnerDistance(int distance) {
+        return cars.stream()
+                .noneMatch(car -> car
+                        .isDistanceEqualTo(distance));
```
ㅋㅋ 이런 건 보통 붙여쓰긴합니다~

devholic22 replied:
수정했습니다 😅

sosow0212 commented on src/main/java/Cars.java:
```diff
@@ -0,0 +1,31 @@
+                        .isDistanceEqualTo(distance));
+    }
+
+    public void getWinnersName(List<String> winners, int distance) {
```
이 메서드를 보면 외부에서 `승자의 점수`를 받아서 매개변수로 넘겨주고, 일급컬렉션 List<Car> 내부에서 하나씩 이 승자의 점수와 같은지 판단하고 있습니다.

즉 결국 getter.equals()와 비슷한 로직을 띄고 있는데요.

이렇게 외부에 의존하는 방식보다는 일급컬렉션 자체적으로 내부 객체를 탐색하면서 우승 점수를 구하고, 우승자까지 구하는 방법은 어떨까요?

내부 객체를 관리할 수 있다는 것이 일급컬렉션의 장점인데, 이 장점을 외부에서 값으로 주입받아서 처리하는 방법은 다소 아쉬워보여요.
물론 처음이라 익숙치 않겠지만, 시도해보시면 좋을 것 같습니다!

devholic22 replied:
와.. 생각해보니 `List<String> winners`가 아예 필요하지 않은 방식이었는데 외부에게 의존되도록 만들었었네요 감사합니다! 반영하였습니다 :)

devholic22 replied:
`내부 객체를 탐색하면서 우승 점수를 구하고` 이 부분을 하려면, 기존에 개발된 `Car from(Name name)`을 `Car of(Name name, int goal)`로 만들어야 할 것 같은데 이런 식으로 진행해보면 될까요? 이 부분에 대해서는 질문 먼저 드려보고 싶습니다..!

sosow0212 commented on src/main/java/DiceMovableImpl.java:
```diff
@@ -0,0 +1,9 @@
+public class DiceMovableImpl implements Dice {
```
전략패턴 잘 사용하셨네요~
이 객체는 test 패키지 안에 있는게 좋을 것 같습니다.

devholic22 replied:
수정했습니다!

sosow0212 commented on src/main/java/GameController.java:
```diff
@@ -0,0 +1,46 @@
+
+public class GameController {
+
+    private final BufferedReader inputReader = new BufferedReader(new InputStreamReader(System.in));
```
여기도 개행 추가해주세요~

그리고 `BufferedReader`를 선언과 동시에 초기화 한 이유가 있을까요?
생성자로 받아보는 건 어떨까요? (입력 전략이 바뀐다면?)

예시가 버퍼리더여서 와닿지 않을 수 있겠지만, 스프링의 DI 개념과 엮어서 생각해보시면 조금 더 와닿으실거같습니다!

devholic22 replied:
실제 스프링에서 수동으로 빈을 등록하여 주입시켜주는 방식을 모방하여, Application 실행 시 BufferedReader를 만들고 GameController에게 인자로 넘겨주는 방식으로 개선해봤습니다! 조언 감사합니다 :)

sosow0212 commented on src/main/java/Name.java:
```diff
@@ -0,0 +1,56 @@
+        return name.length() >= MINIMUM_CAR_NAME_LENGTH && name.length() <= MAXIMUM_CAR_NAME_LENGTH;
+    }
+
+    private static Name convertStringToName(String name) {
```
`public static Name from(String name)`이 메서드와 중복으로 보입니다

devholic22 replied:
와 진짜 너무 무의미했던 메서드였네요.. 삭제 조치 했습니다 감사합니다!

sosow0212 commented on src/main/java/Printer.java:
```diff
@@ -0,0 +1,49 @@
+        System.out.println("실행 결과");
+    }
+
+    public static void alertWinners(Cars playCars, int playDistance) {
```
view에서 도메인 내부까지 다 알아야할까요?

이렇게 생각해볼까요?
만약 View만 담당하는 작업자가 있고(프론트엔드), Domain만 담당하는 작업자가 있습니다.(백엔드)

프론트엔드 개발자가 백엔드 개발자가 개발한 도메인 로직을 모두 다 알아야할까요?

이런 이유로 우리도 스프링 REST API를 개발하면 DTO로 필요한 것만 주고 있습니다.
현재는 DTO를 사용하지 말고 값만 전달해서 받는 건 어떨까요?

devholic22 replied:
`List<String> winners`를 사용하도록 변경하였습니다!

sosow0212 commented on src/test/java/CarTest.java:
```diff
@@ -0,0 +1,49 @@
+
+public class CarTest {
+
+    @Test
```
이것도 사바사이긴 하지만 보통 중요한 어노테이션일 수록 메서드와 가깝습니다!
그래서 저같은 경우는 두 어노테이션의 순서를 바꾸고 있습니다.

devholic22 replied:
말씀해주신 테스트 이름 관례에 따라 `DisplayName`을 아예 삭제하는 식으로 교체했습니다!

sosow0212 commented on src/test/java/CarTest.java:
```diff
@@ -0,0 +1,49 @@
+        List<Name> names = Name.createNameList(nameInput);
+
+        // when
+        List<Car> cars = names.stream().map(Car::from).collect(Collectors.toList());
```
이 부분도 Cars 객체에서 자동으로 내부적인 로직으로 만들 수 있는 정팩메 같은 것을 구현한다면
굳이 스트림으로 로직을 직접 안 만들어도 되겠죠?

테스트에서 예시 
```java
// when
Cars cars = Cars.from(List<Names> names);
```

devholic22 replied:
구현 완료했습니다 :)

sosow0212 commented on src/test/java/DiceTest.java:
```diff
@@ -0,0 +1,51 @@
+        // given
+        Name name = Name.from("hello");
+        Car car = Car.from(name);
+        Dice movableDice = new DiceMovableImpl();
```
전략패턴 👍

sosow0212 commented on src/test/java/NameTest.java:
```diff
@@ -0,0 +1,100 @@
+
+    @Test
+    @DisplayName("정상적인 이름 리스트 입력 시 모두 저장되어야 한다.")
+    public void createNameListTest() {
```
추가적으로 테스트 이름 짓는 방법도 여러가지 방법이 있습니다. 
어차피 테스트 클래스의 메서드이므로 끝에 Test는 안 붙여도 될 것 같아요~

1. @DisplayName으로 한글 설명 + 테스트 메서드명은 영어로 짓기
1-1. 영어로 지을 땐 ex. save_all_when_~~
1-2. 다른 방식으로는 ex. saveAllWhen ~~

2. @DisplayName 사용하지 않고 한글로 테스트명 짓기
- public void 이름_리스트_~~

devholic22 replied:
수정했습니다 테스트 이름에 대한 방법도 알게 되네요! 감사합니다 :)

sosow0212 commented on src/test/java/NameTest.java:
```diff
@@ -0,0 +1,100 @@
+        String nameInput = null;
+
+        // when & then
+        assertThrows(NullPointerException.class, () -> {
```
예외 메서드는
`assertThatThrownBy()` 사용해보는 것도 좋은 것 같습니다.
예외 발생시 어떤 클래스가 나오는지는 물론이고 메시지 등등 여러가지를 검증할 수 있습니다.

devholic22 replied:
말씀하신 것을 보고 [테스트 예외 자료](https://v3.leedo.me/devs/68)를 찾아보니, 테스트 코드 예외 작성 방법이 JUnit5와 AssertJ 라이브러리를 활용한 방식으로 나눠져 있다는 것을 알았습니다 :) 왜 예외 메서드를 이것들로 했는지 물어볼 수도 있겠네요! 각 방식의 특징과 사용법에 대해 더 익혀봐야겠다고 생각이 들었고, 특히 후자 방식은 체이닝이 가능하다는 것도 신기하네요 😲 감사합니다!

sosow0212 commented on src/test/java/CarTest.java:
```diff
@@ -0,0 +1,49 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class CarTest {
```
Cars 테스트는 빠진 것 같습니다!

devholic22 replied:
반영하였습니다 :)

sosow0212 reviewed (changes requested):
잘 반영해주셨습니다~
전략패턴도 잘 사용해주셨고 좋습니다.
몇 부분 수정하면서 패키지도 MVC 구조로 개선해보는 건 어떨까요?

devholic22 committed:
refactor: 상수와 객체 간 공백 추가

devholic22 committed:
refactor: car.isDistanceEqualTo 붙여쓰기

devholic22 committed:
refactor: Dice 패키지 위치 변경 (테스트 구현)

devholic22 committed:
refactor: convertStringToName 메서드 삭제

devholic22 committed:
refactor: 테스트 메서드의 Test 접미사 삭제

devholic22 committed:
refactor: BufferedReader 주입받도록 변경 조치

devholic22 committed:
refactor: 테스트 예외 시 assertThrows 말고 assertThatThrownBy 사용

devholic22 committed:
refactor: DisplayName 삭제

devholic22 committed:
refactor: 일급 컬렉션 장점 활용하도록 변경 (getWinnersName)

devholic22 committed:
refactor: View가 도메인을 알지 못하도록 변경

devholic22 committed:
refactor: Car가 Printer를 알지 못하도록 수정

devholic22 committed:
refactor: Car 생성자 수정

devholic22 committed:
refactor: Car static 메서드 일반 메서드로 변경

devholic22 committed:
refactor: Cars 정적 팩터리 메서드 생성

devholic22 committed:
refactor: CarsTest 생성

devholic22 committed:
refactor: Cars에서 dice를 처음에 초기화하여 사용하도록 변경

devholic22 committed:
test: 처음에는 승자가 없어야 한다.

sosow0212 commented on src/main/java/Application.java:
```diff
@@ -1,9 +1,12 @@
     public static void main(String[] args) throws IOException {
-        GameController game = new GameController();
+        BufferedReader inputReader = new BufferedReader(new InputStreamReader(System.in));
+        GameController game = new GameController(inputReader);
```
굿 좋습니다 외부에서 주입해줄 때 장점은 테스트도 유연하게 할 수 있다는 점이 있습니다.

만약 클래스를 생성할 때 내부적으로 랜덤값을 생성해주는 클래스를 내부적으로 생성할 때 (해당 클래스는 인터페이스 구현체), 테스트를 할 때 랜덤한 값으로 인해 테스트의 실패와 성공이 확률적으로 될 수 있습니다.

만약 외부에서 생성자를 통해 이 랜덤 값을 만들어주는 인터페이스의 구현체를 넣어준다면 어떨까요?
테스트에서는 성공만 하는 구현체를 넣어서 의도한대로 테스트 할 수 있겠죠 (이를 전략패턴이라고 합니다.)

스프링에서도 마찬가지인 이유로 생성자 주입을 권장하고 있습니다!

sosow0212 commented on src/main/java/Car.java:
```diff
@@ -3,19 +3,20 @@ public class Car {
+        this.distance = distance;
     }
 
     public static Car from(Name name) {
```
해당 경우와 같이 리팩토링 된 경우 from보다는 createDefault와 같은 네이밍이 조금 더 좋을 것 같네요~

devholic22 replied:
반영하였습니다 :)

sosow0212 commented on src/main/java/Cars.java:
```diff
@@ -3,29 +3,41 @@
 
-    public void getWinnersName(List<String> winners, int distance) {
-        List<String> winnerNames = cars.stream()
+    public List<String> getWinnersName(int distance) {
```
어떤 부분에서는 메서드 파라미터 변수 앞에 final을 붙이고 안 붙이기도 하는데 기준이 있을까요?

devholic22 replied:
아예 [final 키워드를 붙이는 게 컨벤션으로 정해진 경우](https://hudi.blog/intellij-final-keyword/)도 있음을 확인하였습니다! final의 특성을 이용하여 `가급적 불변으로 사용되도록 명시할 때` 쓰이는 것 같습니다. 말씀하신 부분을 보며 아예 이 부분이 인텔리제이에서 설정으로 있다는 것도 알게 되었습니다. 개인적인 생각으로는, 무조건 모든 메서드 인자에 다 final을 붙이는 것 보다는 변경되면 안 되는 상황에서만 쓰이는 게 좋을 것 같습니다! (그렇지만 이 과제를 하면서 이런 부분은 잘 고려하지 못했네요... 😢)

sosow0212 replied:
해당 부분은 개발을 하시다보면 자신의 기준이 생기게 됩니다! 물론 팀마다 컨벤션이 달라서 지켜야하지만, 팀마다 컨벤션이 다르기 때문에 언제 써도 되고 안되는지 판단하는게 중요한 것 같습니다.

sosow0212 commented on src/main/java/Cars.java:
```diff
@@ -3,29 +3,41 @@
     public void racing() {
-        final Dice dice = new DiceImpl();
-        cars.forEach(car -> car.racing(dice));
+        cars.forEach(car -> {
+            car.racing(dice);
+            Printer.printCarDistanceStatus(car.getName(), car.getStartDistance(), car.getDistance());
+        });
```
이 부분을 한 번에 하기 때문에 Printer를 쓸 수 밖에 없게 되어있네요.

해당 메서드는 자동차의 경주도 진행시키고 프린터를 통해 출력도 하고 있는데요.
즉, 두 가지의 역할을 하고 있습니다.

경주만 하는 메서드, 값을 반환해주는 메서드를 각각 구현한다면
뷰와 도메인이 서로 의존하는 것을 막을 수 있을 것 같습니다!

메서드는 적게 쓴다고 꼭 좋은 것만은 아닙니다.
의존이 분리된다면 오히려 메서드를 통해 책임을 분리해주는 것이 더 좋을 수 있습니다 :)

devholic22 replied:
차들의 각 상태를 `HashMap<String, Integer>` 형태로 담아두는 메서드 (`getAllCarsDistanceStatus`)를 Cars 내부에 만들고, 해당 메서드를 GameController에서 Printer와 연결하는 식으로 하여 Printer와 Cars의 직접적인 연결을 끊어내었습니다!

sosow0212 replied:
사실 `Map<Type, Type>` 구조로 보내주는 것도 어떤 값이 있는지 모르고 모호한 면이 있습니다 ㅠ
변수명을 잘 지어준다면 괜찮겠지만, Map이 아닌 Dto도 고려해보는 것도 좋을 것 같습니다.

그 방법이 아니여도 getter()를 이용해서 뜻이 있는 값을 주는 것도 좋아보이지만, 현재로써 Map이 최선이라고 생각한다면 계속 진행하시면 될 것 같습니다!

sosow0212 commented on src/main/java/GameController.java:
```diff
@@ -26,18 +28,15 @@ public void play() throws IOException {
-                .collect(Collectors.toList());
+    private List<Name> createCarNames(String input) {
+        return Name.createNameList(input);
     }
```
Name이 Car안에 있는데, List<Name>을 만들고 이를 통해 List<Car> 객체를 생성하는 이유가 있을까요?

String input을 일급 컬렉션인 Cars의 정적 팩토리 메서드로 넣어주고 객체 안에서 판단해주는 방법은 없을까요?

devholic22 replied:
`Name.createNameList`를 Cars의 정적 팩터리 메서드에 넣어둠으로써 대체했는데, 이렇게만 해도 충분한지 궁금합니다!

sosow0212 replied:
조금 더 좋은 방법이 있긴한데..!
다음 미션에서도 아마 비슷하게 진행하니 그 때 더 다루도록 할게요~

sosow0212 commented on src/test/java/CarsTest.java:
```diff
@@ -0,0 +1,34 @@
+        Cars cars = Cars.from(names);
+
+        // then
+        Assertions.assertThat(cars.getCarListCount()).isEqualTo(2);
```
static import 해주세요~

devholic22 replied:
[해당 글](https://velog.io/@kasania/Java-Static-import%EC%97%90-%EB%8C%80%ED%95%9C-%EA%B4%80%EC%B0%B0)을 보고 static import를 해야 하는 이유를 알 수 있었네요 😲 적용 완료하였습니다 :)

sosow0212 commented on src/test/java/CarsTest.java:
```diff
@@ -0,0 +1,34 @@
+    public void 이름이_잘못되었다면_자동차_자체가_생성이_되지_않아야_한다() {
+        // given
+        String nameInput = "hello, springboot, test";
+        List<Name> names = Name.createNameList(nameInput);
+
+        // when
+        Cars cars = Cars.from(names);
```
위에서도 말했지만 Cars.from(final String nameInput)을 해버리면 테스트 코드도 간결해지겠죠?

```
// given
String nameInput = "...";

// when
Cars cars = Cars.from(nameInput);
```

이런 식으로요

devholic22 replied:
반영하였습니다 :)

sosow0212 commented on src/test/java/CarsTest.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    @Test
+    public void 처음에는_승자가_없어야_한다() {
```
테스트가 꼼꼼하면 좋긴 하지만 해당 테스트를 만드신 이유가 있을까요?

devholic22 replied:
음.. 사실 테스트 메서드를 생각할 때, 생각할 수 있는 경우들을 고려하여 작성했었습니다. 그래서 그런지 당연할 것 같은 상황에서도 테스트를 작성하였는데, `테스트 할 메서드들을 떠올리는 방법`에 대해서 여쭤봐도 될까요? 뭔가 일관된 저만의 규칙이 없고 단지 떠올리는 대로 작성해둔 것 같아 이 점에 대해 문의드리고 싶습니다.

sosow0212 replied:
사실 저는 단순 getter 같은 메서드는 잘 검증하지 않긴 합니다~
반면에 비즈니스 로직이 있는 그런 메서드는 필수적으로 테스트를 작성하고 있습니다.
이 부분도 다음 미션 진행하면서 더 얘기 나눠보도록해요~

sosow0212 commented on src/main/java/Cars.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    public List<String> getWinnersName(int distance) {
+
```
불필요한 개행 삭제해주세요

devholic22 replied:
반영했습니다 :)

sosow0212 reviewed (changes requested):
처음에 비해 많이 좋아진게 보입니다!

몇 부분에 대한 코멘트 확인해주세요.
해당 부분만 고친다면 아마 마지막 리뷰가 될 것 같습니다.

처음이라 많이 어려우셨겠지만, 조금만 하다보면 금방 익숙해지고, 오히려 예전 코드를 보면 왜 그렇게 짰는지 생각도 듭니다 ㅋㅋ

나머지는 다음 미션에서 진행해보면 좋을 것 같아요.
뷰랑 도메인 의존하는 것만 해결하면 될 것 같습니다!

devholic22 committed:
refactor: 불필요한 개행 삭제

devholic22 committed:
refactor: Assertions static import

devholic22 committed:
refactor: Car 정적 팩터리 메서드명 변경

devholic22 committed:
refactor: Cars 정적 팩터리 메서드에 String 넣도록 함

devholic22 committed:
refactor: 경주 메서드와 값 (상태) 반환 메서드 분리

sosow0212 commented on src/main/java/Cars.java:
```diff
@@ -33,10 +34,17 @@ public List<String> getWinnersName(int distance) {
     }
 
+    public HashMap<String, Integer> getAllCarsDistanceStatus() {
+        HashMap<String, Integer> carsInfo = new HashMap<>();
```
info라는 네이밍은 너무 추상적입니다.

sosow0212 reviewed (approved):
수고하셨습니다!
이번 미션에서 못 다룬 내용은 다음 미션에서 자세하게 다뤄보도록 할게요~

처음 코드에 비해 많이 깔끔해지고 객체지향적으로 변한 것 같아서 좋습니다.
다음 미션에서도 이와 같이 진행해봅시다!

추가적으로 패키지 구조도 다음 미션에서는 나눠주면 좋을 것 같습니다 👍

devholic22 commented:
> 수고하셨습니다! 이번 미션에서 못 다룬 내용은 다음 미션에서 자세하게 다뤄보도록 할게요~
> 
> 처음 코드에 비해 많이 깔끔해지고 객체지향적으로 변한 것 같아서 좋습니다. 다음 미션에서도 이와 같이 진행해봅시다!
> 
> 추가적으로 패키지 구조도 다음 미션에서는 나눠주면 좋을 것 같습니다 👍

넵 감사합니다! 확실히 미션을 진행해보니 너무 일관된 규칙 없이 막 코드를 설계했다는 게 느껴지네요.. 다음 미션도 열심히 해보겠습니다 🔥
